jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/FrameFlusher.java

```diff
@@ -259,26 +276,47 @@ private int getQueueSize()
         }
     }
 
-    @Override
-    public void succeeded()
-    {
-        succeedEntries();
-        super.succeeded();
-    }
-
-    private boolean succeedEntries()
+    public void timeoutExpired()
     {
-        boolean hadEntries = false;
-        for (Entry entry : entries)
+        boolean failed = false;
+        synchronized (FrameFlusher.this)
         {
-            hadEntries = true;
-            if (entry.frame.getOpCode() == OpCode.CLOSE)
-                endPoint.shutdownOutput();
-            notifyCallbackSuccess(entry.callback);
-            entry.release();
+            if (closedCause != null)
+                return;
+
+            long time = System.currentTimeMillis();
+            long nextTimeout = -1;
+
+            Iterator<Entry> iterator = TypeUtil.concat(entries.iterator(), queue.iterator());
+
+            while (iterator.hasNext())
+            {
+                Entry entry = iterator.next();
+                long timeSinceCreation = time - entry.getTimeOfCreation();
+
+                if (timeSinceCreation >= idleTimeout)
+                {
+                    LOG.warn("FrameFlusher write timeout on entry: {}", entry);
+                    failed = true;
+                    canEnqueue = false;
+                    closedCause = new WebSocketWriteTimeoutException("FrameFlusher Write Timeout");
+                    failedEntries.addAll(entries);
+                    failedEntries.addAll(queue);
+                    entries.clear();
+                    queue.clear();
+                    break;
+                }
```

Comment:
else if not a ping/pong we can break the loop as an optimisation.  yes???

Reply:
I'm not really convinced that this is the case

1. enqueue a ping with time stamp 10ms
2. that ping is then polled into entries queue
3. enqueue another ping frame at time 15ms
4. enqueue a standard frame at time 20ms

so then the iterator looks like 
ping(15ms) -> frame(20ms) -> ping(10ms)

and here that optimisation doesn't work?

Reply:
Ah OK, the pings only jump the first queue but not the inner queue.  Doh!